File: dfs/剑指 Offer 12. 矩阵中的路径.go
package main

import "fmt"

func exist(board [][]byte, word string) bool {
	var visited [][]string = make([][]string, len(board))

	for idx, _ := range visited {
		var temp []string = make([]string, len(board[0]))
		for j, _ := range temp {
			temp[j] = "unvisited"
		}
		visited[idx] = temp
	}
	for row := 0; row < len(board); row++ {
		for col := 0; col < len(board[0]); col++ {
			if word[0] == board[row][col] {
				visited[row][col] = "visited"
				if dfs(visited, board, word, 1, row, col) {
					return true
				}
				visited[row][col] = "unvisited"
			}
		}
	}
	return false
}

var movement [][]int = [][]int{
	{1, 0},
	{-1, 0},
	{0, 1},
	{0, -1},
}

func dfs(visited [][]string, board [][]byte, word string, word_idx int, cur_row int, cur_col int) bool {
	if word_idx >= len(word) {
		return true
	}
	for _, val := range movement {
		x := cur_row + val[0]
		y := cur_col + val[1]

		if (x < 0 || x >= len(board)) || (y < 0 || y >= len(board[0])) {
			continue
		}
		if visited[x][y] == "visited" {
			continue
		}
		if word[word_idx] == board[x][y] {
			visited[x][y] = "visited"
			res := dfs(visited, board, word, word_idx+1, x, y)
			visited[x][y] = "unvisited"
			if res {
				return true
			}
		}

	}
	return false
}

func main() {
	board := [][]byte{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'F'}}
	word := "ABCB"
	fmt.Println(exist(board, word))
}

File: dfs/剑指 Offer 12. 矩阵中的路径.py
import copy


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if len(word) == 1:
            for x in range(len(board)):
                for y in range(len(board[0])):
                    if board[x][y] == word[0]:
                        return True
            return False

        visited = copy.deepcopy(board)

        def dfs(word_idx: int, cur_board_row: int, cur_board_col: int, ):
            for delta_x, delta_y in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                if cur_board_row + delta_x < 0 or cur_board_row + delta_x >= len(board):
                    continue
                if cur_board_col + delta_y < 0 or cur_board_col + delta_y >= len(board[0]):
                    continue
                x = cur_board_row + delta_x
                y = cur_board_col + delta_y
                if visited[x][y] == 'visited':
                    continue
                if word[word_idx] == board[x][y]:
                    if word_idx == len(word) - 1:
                        return True
                    else:
                        visited[x][y] = 'visited'
                        if dfs(word_idx + 1, x, y, ):
                            visited[x][y] = 'unvisited'
                            return True
                        visited[x][y] = 'unvisited'
            return False

        for x in range(len(board)):
            for y in range(len(board[0])):
                if board[x][y] == word[0]:
                    visited[x][y] = 'visited'
                    if dfs(1, x, y):
                        return True
                    visited[x][y] = 'unvisited'
        return False

